oci: fix a performance regression from execs

haircommander opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/master/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

<!--
/kind api-change
/kind bug
/kind ci
/kind cleanup
/kind dependency-change
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake
/kind other
-->
/kind bug
#### What this PR does / why we need it:
turns out, starting a new inotify watcher for each exec probe was really inefficient. Replace it with one single watcher for all execs, which makes the RSS addition per exec much less (pretty much the overhead of exec.Cmd, which is unavoidable)
#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
fix a performance regression with exec probes
```

haircommander commented:
/test e2e-gcp

haircommander commented:
/test e2e-gcp

TomSweeneyRedHat commented:
Just curious why you're creating notify in CRI-O rather than fixing upstream?
Otherwise, LGTM once the last unhappy test gets happy.

haircommander commented:
> Just curious why you're creating notify in CRI-O rather than fixing upstream?
> Otherwise, LGTM once the last unhappy test gets happy.

I'm not super sure which upstream you're talking about. I would really like to get a timeout feature into runc, but there is a lot of nuance to the behavior here (if exec takes more than x seconds, sig kill it is not a very generic requirement). So we do the next best thing: our best in CRI-O.

TomSweeneyRedHat commented:
Maybe I misunderstood this part, but it looked like you dropped github.com/rjeczalik/notify and replaced it with your own spin on notify.

mrunalp commented on internal/oci/runtime_oci.go:
```diff
@@ -1101,29 +1101,38 @@ func (c *Container) conmonPidFilePath() string {
-		return nil, err
-	}
 	go func() {
+		defer watcher.Close()
```
This could be put after the err check on NewWatcher.

haircommander replied:
I leave it here because there exists the chance we'll want to add a done channel in this goroutine (it does not seem to be needed now). Leaving it in the goroutine allows us just to extend the for loop, not move around the channels. If that isn't clear though I could document/change it. wydt

mrunalp commented on internal/oci/runtime_oci.go:
```diff
@@ -1101,29 +1101,38 @@ func (c *Container) conmonPidFilePath() string {
-		defer close(eiCh)
-		defer notify.Stop(eiCh)
 		for {
 			select {
```
Comparing with the container exit monitor, we are missing a case for wacher.Errors.

saschagrunert commented on internal/oci/runtime_oci.go:
```diff
@@ -285,13 +286,7 @@ func (r *runtimeOCI) ExecSyncContainer(ctx context.Context, c *Container, comman
-	defer os.RemoveAll(pidDir)
-
-	pidFile := filepath.Join(pidDir, c.id)
+	pidFile := filepath.Join(r.execNotifier.Directory(), c.id+uuid.New().String())
```
What about using `ioutil.TempFile` together with the `r.execNotifier.Directory`?
```suggestion
	pidFile, err := ioutil.TempFile(r.execNotifier.Directory(), "exec-")
```

haircommander replied:
ioutil.TempFile creates the file, which interacts strangely with the inotify code. I think it's better staying as is

saschagrunert reviewed (commented):
Just a nit, otherwise LGTM

haircommander commented:
> Maybe I misunderstood this part, but it looked like you dropped github.com/rjeczalik/notify and replaced it with your own spin on notify.

ah not quite. I introduced a second library in the initial commit because it seemed to work better. I reverted that addition in this PR

haircommander commented:
/retest

mrunalp commented:
/lgtm

saschagrunert reviewed (approved):
/lgtm
/retest
